ILLink: Sweep .override for interface method if the .interfaceImpl is removed

jtschuster opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/102796.

When static interface methods are kept but an implementation of the method is only accessed via a direct call on the implementing type, the implementation method and the .override pointing to the interface method are both kept, but not the .interfaceImpl. This causes a TypeLoadException when the .override points to a method on an interface that the type doesn't implement. To fix this, we needed to update the condition for sweeping overrides to include the case where both the interface and the interface method are kept, but the .interfaceImpl is not kept.

Tests are for static interface methods, but concrete and generic to test type resolution in SweepStep, which can be finicky. Instance methods shouldn't hit this issue since public interface methods don't have a .override and private methods can't be marked without instantiating the type or reflecting over the type, both of which mark the type's .interfaceImpls, and making the .override valid.

jtschuster committed:
Fix issue with static interface methods when interface implementation is removed

jtschuster committed:
Fix type name in string after rename

agocke commented:
>  Instance methods shouldn't hit this issue since public interface methods don't have a .override

Is that required by the ECMA specification? I believe Roslyn can generate these cases, for instance:

```C#
interface I {
 void Method();
}
class A { }

class B : A, I 
{
    void I.Method() { }
}
class C : B 
{
    public void Method() { }
}
class D : C, I { }
```

In general, I would say that it is too hard to guess at what Roslyn will or won't generate -- it would require a completely up-to-date, encyclopedic knowledge of the compiler. We should instead follow the ECMA specification.

agocke commented:
Never mind, the above example doesn't work -- Roslyn generates a private stub, which has the methodimpl. However, I think the point is the same -- I'm not confident there's _no_ situation that Roslyn won't generate this IL. And it looks perfectly valid from the spec perspective, to me.

jtschuster commented on src/tools/illink/src/linker/Linker.Steps/SweepStep.cs:
```diff
@@ -456,21 +456,38 @@ void SweepOverrides (MethodDefinition method)
+		private bool IsInterfaceImplemented (TypeDefinition type, TypeDefinition iface)
+		{
+			var allInterfaces = type.GetAllInterfaceImplementations (Context, false);
+			foreach(var impl in allInterfaces) {
```
I need to update my VS settings to format on save 🙃

agocke replied:
lol I just thought this was another weird Mono formatting rule

jtschuster commented on src/tools/illink/src/linker/Linker.Steps/SweepStep.cs:
```diff
@@ -456,21 +456,38 @@ void SweepOverrides (MethodDefinition method)
+			var allInterfaces = type.GetAllInterfaceImplementations (Context, false);
+			foreach(var impl in allInterfaces) {
+				// Using the resolve cache here is okay, if we resolve to a removed type than the impl will not be marked and we'll continue
+				if (Context.Resolve(impl.InterfaceType) == iface && Context.Annotations.IsMarked (impl)) {
```
Good catch, I think that would cause issues. I'll try doing this recursively like we do in MarkStep.

jtschuster commented:
> > Instance methods shouldn't hit this issue since public interface methods don't have a .override

> Is that required by the ECMA specification? 

Thats a good point, it'll require writing tests in IL, but I think it might be possible to break that assumption.

jtschuster committed:
Use MarkStep's 'is interface implementation marked' logic

jtschuster committed:
Add tests from PR review

- inherited interface methods with overrides
- instance methods on classes with .overrides

jtschuster commented:
I was able to make methods with .overrides on public methods, but I wasn't able to find a way to mark only the method without marking the type as 'RelevantToVariantCasting' or instantiated, so the interfaceImpl is always marked. Am I missing a way to mark the instance method without marking the interfaces too?

jtschuster committed:
Format in conditionally compiled zone

jtschuster committed:
One more format

jtschuster committed:
Mark the correct ifaceImpl, add tests

In MarkMethod, we mark the interfaceImpl that has the same type
definition. If there are multiple implementations of a generic
interface, each with different type parameters, we would only mark the
first such interface. It just so happened that we never hit any issues
because elsewhere we would keep the interface implementations since the
interface definition was marked and the type was 'relevant to variant
casting'.

Adds tests for instance interface methods with .overrides in the
public implementations. It doesn't seems possible to create a scenario
where an instance method could be marked without the declaring type
being marked 'relevant to variant casting' and all the interfaces being
marked. Also, all methods pointed to by a .override is marked by
ProcessMethod, so there isn't a scenario that would have dangling
references in the .override.

jtschuster commented:
For instance methods, we will always mark any methods referenced in a .override as well as their corresponding interface implementation, so we won't end up in a scenario where an instance method would have dangling references in a .override. I did fix up the interface implementation marking (`MarkRuntimeInterfaceImplementation`) to make sure it can find a recursive interface, and marks the .interfaceImpl with the exact same TypeReference rather than just the same TypeDefinition (for example, `IGeneric<int>` should be marked, not just the first `IGeneric<>` interfaceImpl). This was one of the examples where the trimmer would end up doing the right thing (all IGeneric<> implementations would end up marked anyway), but it would do it for a different reason than we expect.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into StaticIfaceIssue

jtschuster committed:
Update test with generics in KeptInterfaceOnTypeInAssembly

MichalStrehovsky reviewed (commented):
I'd leave final review for Sven, but this looks reasonable to me. The interface stripping optimization is difficult to think about and source of many bugs (https://github.com/dotnet/linker/issues/1235, https://github.com/dotnet/linker/issues/2056 off the top of my head, but also this one and probably others). I always wondered how much it actually saves.

It's unfortunate we don't mark the overrides directly in the mark step and instead have to rely on indirect signals to decide whether they should be kept. It's a contrast with how e.g. interface lists are modeled (interface list is modeled in Cecil as a list of InterfaceImplementation (not a list of TypeReference) and we mark individual InterfaceImplementations. Override list is modeled as a list of MethodReference and we don't have a good thing to mark, unless we start marking a tuple `[TypeDefinition, MethodReference]`).

jtschuster commented:
> The interface stripping optimization is difficult to think about and source of many bugs... I always wondered how much it actually saves.

I looked into this and it looks like with the `unusedinterfaces` optimization, the trimmer removes an additional ~13kb from ~2.5mb on a hello world and ~16kb from ~26mb in the ASP.NET Core benchmarks application. Maybe it would be worth disabling by default.

vitek-karas commented:
> Maybe it would be worth disabling by default.

Just an idea - there's other value in this, not just size. This causes failures if somebody incorrectly annotates their code (I'm thinking about the DI problems where this showed up). In a way it helps with making things more trim compatible :-). But maybe it's not worth the pain....

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into StaticIfaceIssue

jtschuster committed:
Check that each interfaceImpl is marked and add tests

jtschuster commented on src/tools/illink/src/linker/Linker.Steps/SweepStep.cs:
```diff
@@ -456,14 +456,19 @@ void SweepOverrides (MethodDefinition method)
+				if (method.Overrides[i].Resolve () is not MethodDefinition ov
+					|| ov.DeclaringType is null
+					|| (IsLinkScope (ov.DeclaringType.Scope) && ShouldRemove (ov))
+					|| (ov.DeclaringType.IsInterface && !MarkStep.IsInterfaceImplementationMarkedRecursively (method.DeclaringType, ov.DeclaringType, Context)))
```
That scenario does as we'd want, keeps M(), but not the .override, and keeps N() and the .override, but I think it does so because all interface implementations for marked interface definitions are marked if the type is passed as a type argument. I think you're pointing out that IsInterfaceImplementationMarkedRecursively only checks if the first interface implementation in the chain is marked, but doesn't check at other levels in the hierarchy. I couldn't find a program that would cause an issue due to that, but I think it makes sense to change regardless. I updated that method to check if each interface implementation is marked all the way up the hierarchy.

jtschuster committed:
NativeAOT now produces this warning since IEnumerable is kept
